src/rust/engine/src/externs/interface.rs

```diff
@@ -566,6 +576,83 @@ py_class!(class PyNailgunServer |py| {
     }
 });
 
+py_class!(class PyNailgunClient |py| {
+  data executor: PyExecutor;
+  data port: u16;
+
+  def execute(&self, py_signal_fn: PyObject, command: String, args: Vec<String>, env: PyDict) -> CPyResult<PyInt> {
+    use nailgun::NailgunClientError;
+
+    let env_list: Vec<(String, String)> = env
+    .items(py)
+    .into_iter()
+    .map(|(k, v): (PyObject, PyObject)| -> Result<(String, String), PyErr> {
+      let k: String = k.extract::<String>(py)?;
+      let v: String = v.extract::<String>(py)?;
+      Ok((k, v))
+    })
+    .collect::<Result<Vec<_>, _>>()?;
+
+    let port = *self.port(py);
+    let executor_ptr = self.executor(py);
+    let python_signal_fn: Value = py_signal_fn.into();
+
+    let (exit_sender, exit_receiver) = oneshot::channel();
+    let nailgun_fut = nailgun::client_execute(
+      port,
+      command,
+      args,
+      env_list,
+      exit_receiver,
+    );
+
+    let exit_code: Result<i32, String> = with_executor(py, executor_ptr, |executor| {
+        let (sender, receiver) = mpsc::channel();
+
+        let _spawned_fut = executor.spawn(async move {
+          let exit_code = nailgun_fut.await;
+          let _ = sender.send(exit_code);
+        });
+
+        let timeout = std::time::Duration::from_millis(50);
+        let output = loop {
+          let event = receiver.recv_timeout(timeout);
+          if let Some(_termination) = maybe_break_execution_loop(&python_signal_fn) {
+            break Err("Quitting because of explicit interrupt".to_string());
+          }
+
+          match event {
+            Ok(res) => break res.map_err(|e| match e {
+              NailgunClientError::PreConnect(s) | NailgunClientError::PostConnect(s) => format!("Nailgun client error: {:?}", s),
+              NailgunClientError::ExplicitQuit => "Explicit quit".to_string(),
+            }),
+            Err(RecvTimeoutError::Timeout) => {
+              continue;
+            }
+            Err(RecvTimeoutError::Disconnected) => {
+              break Err("Disconnected from Nailgun client task".to_string());
+            }
+          }
+        };
+        debug!("Sending message to nailgun client task to exit.");
+        match exit_sender.send(()) {
```

Comment:
It might be useful to add a TODO which wraps up the exit sending here. From looking at the code, it's not immediately clear to me what `exit_sender.send())` does compared to seeing `sender.send(exit_code)` above. Maybe just comments or renaming these variables might be helpful, but it's confusing to see the `exit_sender` not sending the `exit_code`, for example.